osquery/sql/virtual_table.cpp

```diff
@@ -74,6 +73,194 @@ static inline std::string opString(unsigned char op) {
   return "?";
 }
 
+/// A list of tables that come from extensions; it is used to determine which
+/// table can be read/write
+std::vector<std::string> extension_table_list;
+Mutex extension_table_list_mutex;
+
+bool isExtensionTable(const std::string& table_name) {
+  ReadLock lock(extension_table_list_mutex);
+  return (std::find(extension_table_list.begin(),
+                    extension_table_list.end(),
+                    table_name) != extension_table_list.end());
+}
+
+namespace {
+// A map containing an sqlite module object for each virtual table
+std::unordered_map<std::string, struct sqlite3_module> sqlite_module_map;
+Mutex sqlite_module_map_mutex;
+
+bool getColumnValue(std::string& value,
+                    int index,
+                    int argc,
+                    sqlite3_value** argv) {
+  value.clear();
+
+  if (index >= argc) {
+    return false;
+  }
+
+  auto sqlite_value = argv[index];
+  switch (sqlite3_value_type(sqlite_value)) {
+  case SQLITE_INTEGER: {
+    auto temp = sqlite3_value_int64(sqlite_value);
+    value = std::to_string(temp);
+    break;
+  }
+
+  case SQLITE_FLOAT: {
+    auto temp = sqlite3_value_double(sqlite_value);
+    value = std::to_string(temp);
+    break;
+  }
+
+  case SQLITE_BLOB:
+  case SQLITE3_TEXT: {
+    auto data_pointer = sqlite3_value_blob(sqlite_value);
+
+    auto buffer_size = sqlite3_value_bytes(sqlite_value);
+    value.resize(buffer_size);
+    std::memcpy(&value[0], data_pointer, buffer_size);
+
+    break;
+  }
+
+  case SQLITE_NULL: {
+    break;
+  }
+
+  default: {
+    VLOG(1) << "Invalid column type";
+    return false;
+  }
+  }
+
+  return true;
+}
+
+// Type-aware serializer to pass parameters between osquery and the extension
+int serializeUpdateParameters(std::string& json_value_array,
+                              int argc,
+                              sqlite3_value** argv,
+                              const TableColumns& columnDescriptors) {
+  if (columnDescriptors.size() != static_cast<size_t>(argc - 2)) {
```

Comment:
It could be problematic to substitute something from number and cast it to `unsigned` type. Is here any possibility for `argc` to be less than `2`?

Reply:
All the argc parameters are defined as `int` to match the type used by sqlite3 to define that variable. 

If we switch the type to something else, we will have to cast it to `size_t` when we receive it and back to `int` when we return it back to the library (would you prefer this approach?).

Parameter serialization only happens when an UPDATE or INSERT statement has been received.

This is a short explanation of how it works for this case, but the documentation goes in more detail:

### Update queries
We are updating an existing row; the ID for that row is inside argv[0]

##### argv[1] == NULL
The table code is expected to keep using the ID associated with the row being modified. This ID can be found in argv[0].

##### argv[1] == INTEGER
The table handler is expected to update the row, invalidate the existing row ID (argv[0]) and use the new one found in argv[1].

### Insert queries
In this case, the row is new so there is no existing row ID. This means that argv[0] will be set to SQLITE_NULL.

##### argv[1] == NULL
The table is expected to come up with a new row ID and return it back to sqlite.

##### argv[1] == INTEGER
sqlite requires the table logic to use the supplied row ID when saving the row.

Reply:
Thank you for the explanation. I see your point. But even so I'd like to suggest you to use a `std::size_t` here.
1. `int` in `sqlite3` interface is just a part of interface. And is good to see this conversion from `std::size_t` to `int` only when we use the `sqlite` library.
2. Mixing up a signed and unsigned types is bad idea. It could be cause for many bugs and even UB, which is pure evil. Look at the example:
```c++
#include <iostream>
int main() {
  const auto sv = -2;
  const auto usv = 1u;
  auto res = sv + usv;
  std::cout << res << '\n';
  return 0;
}
```
What is the type of `res` and what is the answer? This is potential UB.
3. You right for high load pieces of code static conversion from one type to another could be have a notable performance penalty. But we can not solve this problem by using a `int` type here. Because there still will be a lot of conversions. And, that is the worst, they will be implicit. Which is hard to find, to control, test and debug.
4. And semantics, `std::size_t` is a special type dedicated to represent indexes, address offsets and so on. If you see that type you will know - that variable is for the size. `int` is too common, variable of this type could be anything.